rir/src/ir/Compiler.cpp

```diff
@@ -145,6 +145,141 @@ Code* compilePromise(CompilerContext& ctx, SEXP exp);
 void compileExpr(CompilerContext& ctx, SEXP exp);
 void compileCall(CompilerContext& ctx, SEXP ast, SEXP fun, SEXP args);
 
+void compileWhile(CompilerContext& ctx, std::function<void()> compileCond,
+                  std::function<void()> compileBody) {
+    CodeStream& cs = ctx.cs();
+
+    BC::Label loopBranch = cs.mkLabel();
+    BC::Label nextBranch = cs.mkLabel();
+
+    ctx.pushLoop(loopBranch, nextBranch);
+
+    unsigned beginLoopPos = cs.currentPos();
+
+    cs << BC::beginloop(nextBranch) << loopBranch;
+
+    compileCond();
+    cs << BC::brfalse(nextBranch);
+
+    compileBody();
+    cs << BC::pop() << BC::br(loopBranch) << nextBranch;
+
+    if (ctx.loopNeedsContext()) {
+        cs << BC::endloop();
+    } else {
+        cs.remove(beginLoopPos);
+    }
+
+    cs << BC::push(R_NilValue) << BC::invisible();
+
+    ctx.popLoop();
+}
+
+bool compileSimpleFor(CompilerContext& ctx, SEXP sym, SEXP seq, SEXP body) {
+    Match(seq) {
+        Case(LANGSXP, fun, argsSexp) {
+            RList args(argsSexp);
+            if (fun != symbol::Colon || args.length() != 2) {
+                return false;
+            }
+
+            SEXP start = args[0];
+            SEXP end = args[1];
+            if (TYPEOF(start) != INTSXP && TYPEOF(start) != LGLSXP &&
+                (TYPEOF(start) != REALSXP || XLENGTH(start) == 0 ||
+                 *REAL(start) != (int)*REAL(start))) {
+                return false;
+            }
+
+            // for(i in m:n) {
+            //   ...
+            // }
+            // =>
+            // i' <- m
+            // n' <- n
+            // if (i' > n') {
+            //   n' <- ceil(n') - 1
+            //   while (i' > n') {
+            //     i <- i'
+            //     i' <- i' - 1
+            //     ...
+            //   }
+            // } else {
+            //   n' <- floor(n') + 1
+            //   while (i' < n') {
+            //     i <- i'
+            //     i' <- i' + 1
+            //     ...
+            //   }
+            // }
+
+            CodeStream& cs = ctx.cs();
+            BC::Label fwdBranch = cs.mkLabel();
+            BC::Label endBranch = cs.mkLabel();
+
+            // i' <- m
+            compileExpr(ctx, start);
+            cs << BC::floor();
```

Comment:
missing ensureNamed

Reply:
Not sure, if `start` is a constant can't it be modified in place? If it's a variable it should already be named and then `floor` will copy

Reply:
yes, and the copy will have refcount of 0. an you use this value multiple times

Reply:
please be a bit careful with the refcounts in general. those are not fun bugs to debug and it's easy to get an error to accidentally pass tests.